File: src/builder.py
# ============================================================
#						 builder.py
# ============================================================
import os
import time
import shutil

from native import *
from variables import *
from core_utils import *
from build_utils import *

def MakeFileSystem():
	if check_dir(output['dir']):
		VexelLog("Cleaning...", "INFO")
		rmdir(output['dir'])
	
	mkdir(output['res'])
	mkdir(output['dex'])
	mkdir(output['classes'])
	mkdir(output['gen'])
	mkdir(output['key'])
	mkdir(output['apk'])


def build(dir=None):
	start = time.time()
	
	config = load_config(project['vexel_build'])
	MakeFileSystem()
	update_manifest(project['manifest'], output['manifest'], config)
	VexelLog("Compile resources...", "BUILD")
	run_aapt2_compile()

	VexelLog("Link resources...", "BUILD")
	run_aapt2_link()
	
	VexelLog("Compile Java...", "BUILD")
	run_java_compile()
	
	# Build native
	build_native()

	VexelLog("Convert to DEX...", "BUILD")
	convert_dex()

	VexelLog("Add classes.dex...", "BUILD")
	shutil.copy(path(output['apk'], "base.apk"), path(output['apk'], "unaligned.apk"))
	
	# Add Dex
	for dex in find_files(output['dex'], ".dex"):
		add_to_apk(path(output['apk'], "unaligned.apk"), dex, os.path.basename(dex), False, 0)
	
	# Add User Dynamic Libraries
	add_to_apk(path(output['apk'], "unaligned.apk"), project['dyn_lib'], "lib")
	
	# Add Compiled Dynamic Libraries
	add_to_apk(path(output['apk'], "unaligned.apk"), output['dyn_lib'], "lib")
	
	# Add Java Libraries
	add_to_apk(path(output['apk'], "unaligned.apk"), project['assets'], "assets")
	
	VexelLog("Align APK...", "BUILD")
	align_apk()

	VexelLog("Sign APK...", "BUILD")
	sign_apk(config)
	
	mkdir(output['final'])
	shutil.copy(path(output['apk'], "final.apk"), path(output['final'], "Debug.apk"))
	
	elapsed = round(time.time() - start, 2)
	VexelLog(f"Build Successful in {elapsed} sec", "INFO")
	VexelLog(f"Apk saved at: {path(output['final'], 'Debug.apk')}", "INFO")

File: src/Builder.kt
// ============================================================
//                         Builder.kt
// ============================================================
package app.pie.vexel

import java.io.File
import java.nio.file.Files
import java.nio.file.StandardCopyOption

fun makeFileSystem() {
    if (checkDir(Vexel.Output.dir)) {
        vexelLog("Cleaning build directory...", "INFO")
        removeDir(Vexel.Output.dir)
    }

    makeDir(Vexel.Output.res)
    makeDir(Vexel.Output.dex)
    makeDir(Vexel.Output.classes)
    makeDir(Vexel.Output.gen)
    makeDir(Vexel.Output.key)
    makeDir(Vexel.Output.apk)
}

fun build() {
    val start = System.currentTimeMillis()

    val config = loadConfig(Vexel.Project.vexelBuild)
	val app = config["app"] ?: mutableMapOf()
	val appName = app.get("name")
	
    makeFileSystem()

    Build.updateManifest(Vexel.Project.manifest, Vexel.Output.manifest, config)

    vexelLog("Compiling resources...", "BUILD")
    Build.runAapt2Compile()

    vexelLog("Linking resources...", "BUILD")
    Build.runAapt2Link()

    vexelLog("Checking Java sources...", "INFO")
    Build.runJavaCompile()
    
    vexelLog("Checking Kotlin sources...", "INFO")
    Build.runKotlinCompile()

    vexelLog("Checking C/C++ sources...", "INFO")
    buildNative()

    vexelLog("Convert classes to DEX...", "BUILD")
    Build.convertDex()

    vexelLog("Add classes.dex...", "BUILD")

    Files.copy(
        File(path(Vexel.Output.apk, "base.apk")).toPath(),
        File(path(Vexel.Output.apk, "unaligned.apk")).toPath(),
        StandardCopyOption.REPLACE_EXISTING
    )

    // Add DEX
    for (dex in findFiles(Vexel.Output.dex, ".dex")) {
        Build.addToZip(path(Vexel.Output.apk, "unaligned.apk"), dex, File(dex).name, false)
    }

    // Add User Dynamic Libraries
	Build.mergeNativeLibraries(
    	Vexel.Project.dynLib,
    	Vexel.Output.dynLib
	)

    // Add Dynamic Libraries
    Build.addToZip(
        path(Vexel.Output.apk, "unaligned.apk"),
        Vexel.Output.dynLib
    )

    // Add Assets
    Build.addToZip(
        path(Vexel.Output.apk, "unaligned.apk"),
        Vexel.Project.assets
    )

    vexelLog("Aligning APK...", "BUILD")
    Build.alignApk()

    Build.signApk()

    makeDir(Vexel.Output.finalApk)

    Files.copy(
        File(path(Vexel.Output.apk, "final.apk")).toPath(),
        File(path(Vexel.Output.finalApk, "${appName}-debug.apk")).toPath(),
        StandardCopyOption.REPLACE_EXISTING
    )

    val elapsed = (System.currentTimeMillis() - start) / 1000.0

    vexelLog("Build Successful in %.2f sec.".format(elapsed), "INFO")
    vexelLog("APK saved at: ${path(Vexel.Output.finalApk,"${appName}-debug.apk")}", "INFO")
}